### ISSUE TITLE
Internal error for NATURAL JOIN and VECTORIZE='experimental_on'

### ISSUE THREAD

**Issue by mrigger** [cockroachdb/cockroach#44547]
Consider the following test case:

```sql
SET SESSION VECTORIZE='experimental_on';
CREATE TABLE t0(c0 INT4);
CREATE TABLE t1(c0 INT8);
INSERT INTO t0(c0) VALUES(0);
INSERT INTO t1(c0) VALUES(0);
SELECT * FROM t0 NATURAL JOIN t1; -- internal error: unexpected error from the vectorized runtime: interface conversion: coldata.column is []int32, not []int64
```

Unexpectedly, the `SELECT` results in an error with the following stacktrace:

```
ERROR: internal error: unexpected error from the vectorized runtime: interface conversion: coldata.column is []int32, not []int64
SQLSTATE: XX000
DETAIL: stack trace:
github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror/error.go:70: func1()
runtime/panic.go:679: gopanic()
runtime/iface.go:255: panicdottypeE()
github.com/cockroachdb/cockroach/pkg/col/coldata/vec.go:202: Int64()
github.com/cockroachdb/cockroach/pkg/sql/colexec/hashtable.eg.go:398: rehash()
github.com/cockroachdb/cockroach/pkg/sql/colexec/hashtable.go:332: computeBuckets()
github.com/cockroachdb/cockroach/pkg/sql/colexec/hashtable.go:363: lookupInitial()
github.com/cockroachdb/cockroach/pkg/sql/colexec/hashjoiner.go:442: exec()
github.com/cockroachdb/cockroach/pkg/sql/colexec/hashjoiner.go:223: Next()
github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:149: next()
github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:140: nextAdapter()
github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror/error.go:91: CatchVectorizedRuntimeError()
github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:177: Next()
github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:170: Run()
github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:749: Run()
github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:370: Run()
github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:375: Run()
github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:977: PlanAndRun()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:877: execWithDistSQLEngine()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:769: dispatchToExecutionEngine()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:471: execStmtInOpenState()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:101: execStmt()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1316: execCmd()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1245: run()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:452: ServeConn()
github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:566: func1()
runtime/asm_amd64.s:1357: goexit()
```

I found this bug based on commit `bf96530216efcaaafcd3e88db22089b4f0269480`.

**Comment by yuzefovich** [comment]
I see a problem - we're assuming that the types in equality columns are exactly the same, and in this case they are not :/

**Comment by yuzefovich** [comment]
The fix here is somewhat involved, so I'll fix it once #44523 merges to avoid merging conflicts.

The root of the problem is that we assume that the types in equality columns are of exactly the same type which is wrong. There is another problem that we assume that the order of columns on both sides of the join is the same which is also not always the case.
